qualang_tools/video_mode/videomode.py

```diff
@@ -0,0 +1,444 @@
+            )
+
+
+class VideoMode:
+    def __init__(
+        self, qm: QuantumMachine, parameters: Dict | ParameterTable, job: QmJob = None
+    ):  # TODO: optional[QmJob] returns an error
+        """
+        This class aims to provide an easy way to update parameters in a QUA program through user input while the
+        program is running. It is particularly useful for calibrating parameters in real time. The user can specify
+        the parameters to be updated and their initial values in the parameter dictionary called ```param_dict```.
+        The video mode will then automatically create the corresponding QUA variables and update them through user
+        input.
+
+        Parameters dictionary should be of the form:
+         ```{"parameter_name": initial_parameter_value }```.
+
+        The way this is done is by adding two methods of this class at the beginning of the QUA program declaration:
+
+        - ```declare_variables```: This method will create the QUA variables corresponding to the parameters to be updated.
+        - ```load_parameters```: This method will load the updated values for the parameters through IO variables  1 and 2.
+        The user can then start the video mode outside the QUA program by calling the ```execute``` method of this class.
+        This will start a new parallel thread in charge of updating the parameters in the parameter table through user input.
+        The user can stop the video mode by entering 'stop' in the terminal.
+        The user can also resume a paused program by entering 'done' again in the terminal.
+
+        Args:
+            qm: Quantum Machine object.
+            parameters: Dictionary or ParameterTable containing the parameters to be updated and their initial values.
+        """
+        self.qm = qm
+        self.job = None
+        self._parameter_table = (
+            parameters
+            if isinstance(parameters, ParameterTable)
+            else ParameterTable(parameters)
+        )
+        self.active = True
+        self.thread = threading.Thread(target=self.update_parameters)
+        self.stop_event = threading.Event()
+        self.implemented_commands = (
+            "List of implemented commands: \n "
+            "get: returns the current value of the parameters. \n "
+            "stop: quit VideoMode. \n "
+            "done: resume program (if pause_program==True). \n "
+            "help: displays the list of available commands. \n "
+            "'param_name'='param_value': sets the parameter to the specified value (ex: V1=0.152).\n "
+            "'param_name': returns the value of the parameter.\n"
+        )
+
+        signal.signal(signal.SIGINT, self.signal_handler)
+        signal.signal(signal.SIGTERM, self.signal_handler)
+
+    def signal_handler(self, signum, frame):
+        """Signal handler for SIGTERM and SIGINT signals."""
+        print(f"Received signal {signum}, stopping VideoMode...")
+        self.qm.set_io1_value(666)
```

Comment:
why 666?

If it's a magic number, then change it from 666, and put it in a constant variable.

Reply:
The value is chosen such that the switch_case that is evaluated within the program could never encounter a case with the input IO value. It could be any large number which would not fit in a reasonable switch case for this particular use case

Reply:
I would go for `-1`, but you can go for `2^31-1`.
Then as I wrote, change it to something nobody will find offending.
In any case, it's a magic number, so put it in a constant variable:
`EMPTY_OPTION = 2**31-1`